packages/substrate-client/src/chainhead/chainhead.ts

```diff
@@ -16,46 +17,43 @@ import {
 } from "@/utils/subscriptions-manager"
 
 function isOpeartionEvent(
-  event: FollowEventWithRuntime | FollowEventWithoutRuntime,
+  event: FollowEventWithRuntime | FollowEventWithoutRuntime | OperationEvents,
 ): event is OperationEvents {
   return (event as OperationEvents).operationId !== undefined
 }
 
-export function follow(
+export function getChainHead(
   request: ClientRequest<
     string,
-    FollowEventWithoutRuntime | FollowEventWithRuntime
+    FollowEventWithoutRuntime | FollowEventWithRuntime | OperationEvents
   >,
-): (
-  withRuntime: true,
-  cb: (event: FollowEventWithRuntime) => void,
-) => FollowResponse
-export function follow(
-  request: ClientRequest<
-    string,
-    FollowEventWithoutRuntime | FollowEventWithRuntime
-  >,
-): (
-  withRuntime: false,
-  cb: (event: FollowEventWithoutRuntime) => void,
-) => FollowResponse
-export function follow(
-  request: ClientRequest<
-    string,
-    FollowEventWithoutRuntime | FollowEventWithRuntime
-  >,
-) {
+): ChainHead {
   return (
     withRuntime: boolean,
     cb:
       | ((event: FollowEventWithoutRuntime) => void)
       | ((event: FollowEventWithRuntime) => void),
   ): FollowResponse => {
     const subscriptions = getSubscriptionsManager()
-    const _genesisHash = new Promise<string>((res) => {
-      request("chainHead_unstable_genesisHash", [], res)
-    })
     let unfollow: () => void = () => {}
+    let firstBlockResolver: (data: string) => void
+    let firstBlockRejecter: (e: any) => void
+
+    const firstBlockPromise = new Promise<string>((res, rej) => {
+      firstBlockResolver = res
+      firstBlockRejecter = rej
+    })
+
+    let _cb = (e: FollowEventWithRuntime) => {
+      if (e.event !== "initialized") {
+        firstBlockRejecter(new Error("Missing 'initialized' event"))
+      } else {
+        firstBlockResolver(e.finalizedBlockHash)
+      }
+      _cb = cb
+      cb(e)
+    }
```

Comment:
We can add [deferred.ts](https://raw.githubusercontent.com/denoland/deno_std/main/async/deferred.ts) to utils folder and write it like this

```ts
const firstBlockPromise = deferred<string>()

let _cb = (e: FollowEventWithRuntime) => {
  if (e.event !== "initialized") {
    firstBlockPromise.reject(new Error("Missing 'initialized' event"))
  } else {
    firstBlockPromise.resolve(e.finalizedBlockHash)
  }
  _cb = cb
  cb(e)
}
```

Reply:
good point! Next time, though... Because I ended up removing it... But yeah, that would have been cleaner, for sure!